tiledb/sm/cpp_api/ndrectangle.h

```diff
@@ -0,0 +1,327 @@
+   * @param dim_idx The index of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  template <class T>
+  NDRectangle& set_range(uint32_t dim_idx, T start, T end) {
+    auto& ctx = ctx_.get();
+    // Create the tiledb_range_t struct and fill it
+    tiledb_range_t range;
+    range.min = static_cast<const void*>(&start);
+    range.min_size = sizeof(T);
+    range.max = static_cast<const void*>(&end);
+    range.max_size = sizeof(T);
+
+    // Pass the struct to tiledb_ndrectangle_set_range
+    ctx.handle_error(tiledb_ndrectangle_set_range(
+        ctx.ptr().get(), ndrect_.get(), dim_idx, &range));
+
+    return *this;
+  }
+
+  /**
+   * Adds a 1D string range along a dimension index, in the form (start, end).
+   * Applicable only to variable-sized dimensions
+   *
+   * @param dim_idx The index of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  NDRectangle& set_range(
+      uint32_t dim_idx, const std::string& start, const std::string& end) {
+    auto& ctx = ctx_.get();
+    // Create the tiledb_range_t struct and fill it
+    tiledb_range_t range;
+    range.min = static_cast<const void*>(start.c_str());
+    range.min_size = start.size();
+    range.max = static_cast<const void*>(end.c_str());
+    range.max_size = end.size();
+
+    // Pass the struct to tiledb_ndrectangle_set_range
+    ctx.handle_error(tiledb_ndrectangle_set_range(
+        ctx.ptr().get(), ndrect_.get(), dim_idx, &range));
+
+    return *this;
+  }
+
+  /**
+   * Adds a 1D string range along a dimension index, in the form (start, end).
+   * Applicable only to variable-sized dimensions
+   *
+   * @param dim_name The name of the dimension to add the range to.
+   * @param start The range start to add.
+   * @param end The range end to add.
+   * @return Reference to this NDRectangle.
+   */
+  NDRectangle& set_range(
+      const std::string& dim_name,
+      const std::string& start,
+      const std::string& end) {
+    auto& ctx = ctx_.get();
+    // Create the tiledb_range_t struct and fill it
+    tiledb_range_t range;
+    range.min = static_cast<const void*>(start.c_str());
+    range.min_size = start.size();
+    range.max = static_cast<const void*>(end.c_str());
+    range.max_size = end.size();
+
+    // Pass the struct to tiledb_ndrectangle_set_range_for_name
+    ctx.handle_error(tiledb_ndrectangle_set_range_for_name(
+        ctx.ptr().get(), ndrect_.get(), dim_name.c_str(), &range));
+
+    return *this;
+  }
+
+  /**
+   * Retrieves a range for a given variable length string dimension index and
+   * range id.
+   *
+   * @param dim_idx The dimension index.
+   * @param range_idx The range index.
+   * @return A pair of the form (start, end).
+   */
+  std::array<std::string, 2> range(unsigned dim_idx) {
```

Comment:
Should this and the others return an `std::tuple` instead of an array?

Reply:
Check https://github.com/TileDB-Inc/TileDB/pull/5056#discussion_r1646402077